Use the same name for team handles in team policies

tretre91 opened the pull request:
This PR aims to improve the usability of team policies in generic contexts by renaming the team member variable for all specializations of team policies to `member`.

Currently, the team member stored in the `*RangeBoundariesStruct` returned by team policies is called `member`, `thread` or `team` depending on the backend, which is not convenient when implementing something that takes any team policy as argument (for context, it is needed for an implementation of local deep copy I am working on).

tretre91 committed:
Use the same name for the team_handle
- Use the variable name "member" for the team handle in the nested
  execution policies

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Replace all occurences of team handles
- replace some occurences of `.team` and `.thread` that were overlooked
  in the last commit

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Rename the team handle for the CudaTask specialization

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 commented:
Retest this please

masterleinad reviewed (approved):
Fine with me.

science-enthusiast commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -403,17 +403,19 @@ struct TeamVectorRangeBoundariesStruct<iType, CudaTeamMember> {
 template <typename iType>
 struct ThreadVectorRangeBoundariesStruct<iType, CudaTeamMember> {
   using index_type = iType;
+  const CudaTeamMember& member;
```
Why is `member` introduced into `ThreadVectorRangeBoundariesStruct ` here, in the case of HIP and in `core/src/Kokkos_ExecPolicy.hpp`?

tretre91 replied:
To be able to have code looking like this:
```cpp
template<class TeamPolicy>
void barrier(const TeamPolicy& policy) {
    policy.member.team_barrier()
}
```
with any team policy and any backend (barriers are not used for `ThreadVectorRange` but it's just to illustrate)

science-enthusiast replied:
Could you share some examples where `member` (could be `CudaTeamMember` or `HIPTeamMember`) can be used and it will be well-defined?

tretre91 replied:
For example, if we need to generate the same policy but with a different count, we need to use the `member` from the old policy, because the TeamPolicy functions always take a team member as 1st arg, even in the case of ThreadVectorRange where the current implementation doesn't store the team member handle.
```cpp
template <typename iTypeOut, typename iType, class TeamMemberType>
Kokkos::Impl::ThreadVectorRangeBoundariesStruct<iTypeOut, TeamMemberType>
    KOKKOS_INLINE_FUNCTION
    local_deep_copy_policy(Kokkos::Impl::ThreadVectorRangeBoundariesStruct<
                               iType, TeamMemberType> const& policy,
                           iTypeOut N) {
  return Kokkos::ThreadVectorRange(policy.member, N);
}
```

science-enthusiast replied:
Thank you for the clarification. Since `member` has not been there before in these structs, it looks like it is going to help the upcoming local deep copy facility. That being the case this PR could focus only on making the names uniform in those places where the member variable is already present but with different names. It is better to introduce this new member variable in those structs, as part of a different PR.

science-enthusiast replied:
The addition of `member` in `ThreadVectorRangeBoundariesStruct` in `core/src/Kokkos_ExecPolicy.hpp` is still retained. But it is ok. It is an incremental change happening at a more abstract level compared to specific backends like CUDA and HIP.

Also, to start with, all the three constructors (in CUDA, HIP and Kokkos_ExecPolicy) have been having an unnamed `TeamMemberType` as the first parameter. Nothing too wrong in actually using the parameter and assigning it to a `member`.

tretre91 replied:
I missed this one, thanks for pointing it out. I will also revert it just to be consistent

science-enthusiast commented on core/src/Cuda/Kokkos_Cuda_Team.hpp:
```diff
@@ -403,17 +403,19 @@ struct TeamVectorRangeBoundariesStruct<iType, CudaTeamMember> {
-  ThreadVectorRangeBoundariesStruct(const CudaTeamMember, index_type arg_begin,
-                                    index_type arg_end)
-      : start(arg_begin), end(arg_end) {}
+  ThreadVectorRangeBoundariesStruct(const CudaTeamMember& thread_,
```
`thread_` could be called `arg_thread` here and in HIP.

Maybe `count` in various places could also be `arg_count`.

tretre91 replied:
Looking at the code, I would maybe rename `arg_begin` and `arg_end` to `begin_` and `end_` to be in line with what is done for TeamVectorRange and TeamThreadRange

science-enthusiast replied:
Sure, @tretre91. Whatever improves consistency.

science-enthusiast replied:
I haven't looked everywhere but I see in `core/src/Cuda/Kokkos_Cuda_Task.hpp` the constructor arguments are named as `arg_...`. Please have a look at more places and decide.

science-enthusiast replied:
This is very approximate. I grepped for `arg_` from the root directory and it looks like naming arguments in constructors with the prefix `arg_` is used a lot in Kokkos.

tretre91 replied:
Ok, I'll look into it

tretre91 committed:
Rename some of the constructor args for ThreadVectorRange

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Revert changes in Cuda and Hip team
- these changes will be applied as part of another pull request

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Use the same names for the constructor arguments
- The names of the constructor arguments for the team policies varied
  between files or sometimes between policies in the same file

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 committed:
Revert the addition of member to the default ThreadVectorRange impl
- this will also be applied in a future PR

Signed-off-by: Trévis Morvany <trevis.morvany@cea.fr>

tretre91 commented:
Retest this please
